core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -412,19 +421,77 @@ public boolean index(Index index) {
+             *
+             *  if A arrives on the shard first we use addDocument since maxUnoptimzeAddDocumentTimestamp is < 10. A` will then just be skipped or calls
+             *  updateDocument.
+             */
+            final boolean skipVersionCheck;
+            if (index.isAutogeneratedID()
+                && index.version() == Versions.NOT_FOUND
+                && index.versionType() == VersionType.INTERNAL) {
+                // this optimization is only enabled for version = -1 && type=internal and nothing else
+                long deOptimizeTimestamp = highestDeOptimzeAddDocumentTimestamp.get();
+                if (index.isCanHaveDuplicates()) {
+                    useAddDocument = false;
+                    do {
+                        if (highestDeOptimzeAddDocumentTimestamp.get() >= index.timestamp()) {
+                            break;
+                        }
+                    } while(highestDeOptimzeAddDocumentTimestamp.compareAndSet(deOptimizeTimestamp, index.timestamp()) == false);
+                    assert highestDeOptimzeAddDocumentTimestamp.get() >= index.timestamp();
+                } else {
+                    useAddDocument = deOptimizeTimestamp < index.timestamp();
+                }
+                skipVersionCheck = useAddDocument == false; // we skip the version check if we can't use the optimization and just
+                // add update the document
+            } else {
+                skipVersionCheck = false;
+                useAddDocument = false;
+            }
+            long currentVersion;
             final boolean deleted;
-            final VersionValue versionValue = versionMap.getUnderLock(index.uid());
-            if (versionValue == null) {
-                currentVersion = loadCurrentVersionFromIndex(index.uid());
-                deleted = currentVersion == Versions.NOT_FOUND;
+            final VersionValue versionValue;
+            if (useAddDocument == false) {
+                versionValue = versionMap.getUnderLock(index.uid());
+                if (versionValue == null) {
+                    currentVersion = loadCurrentVersionFromIndex(index.uid());
+                    deleted = currentVersion == Versions.NOT_FOUND;
+                } else {
+                    currentVersion = checkDeletedAndGCed(versionValue);
+                    deleted = versionValue.delete();
+                }
             } else {
-                currentVersion = checkDeletedAndGCed(versionValue);
-                deleted = versionValue.delete();
+                versionValue = null;
+                currentVersion = Versions.NOT_FOUND;
+                deleted = false;
             }
-
             final long expectedVersion = index.version();
-            if (checkVersionConflict(index, currentVersion, expectedVersion, deleted)) return false;
+            if (skipVersionCheck == false) {
+                if(checkVersionConflict(index, currentVersion, expectedVersion, deleted)) {
```

Comment:
we can't otherwise we might run into version conflicts on retries if the original request make it in first and that would fail the document